src/test/java/net/snowflake/client/jdbc/CompressedStreamFactoryTest.java

```diff
@@ -0,0 +1,94 @@
+package net.snowflake.client.jdbc;
+
+import static org.junit.Assert.assertEquals;
+
+import com.github.luben.zstd.ZstdInputStream;
+import com.github.luben.zstd.ZstdOutputStream;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.zip.GZIPInputStream;
+import java.util.zip.GZIPOutputStream;
+import org.apache.http.Header;
+import org.apache.http.message.BasicHeader;
+import org.junit.Test;
+
+public class CompressedStreamFactoryTest {
+
+  private final CompressedStreamFactory factory = new CompressedStreamFactory();
+
+  @Test
+  public void testDetectContentEncodingAndGetInputStream_Gzip() throws Exception {
+    // Original data to compress and validate
+    String originalData = "Some data in GZIP";
+
+    // Creating encoding header
+    Header encodingHeader = new BasicHeader("Content-Encoding", "gzip");
+
+    // Creating a gzip byte array using GZIPOutputStream
+    byte[] gzipData;
+    try (ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+        GZIPOutputStream gzipOutputStream = new GZIPOutputStream(byteArrayOutputStream)) {
+      gzipOutputStream.write(originalData.getBytes(StandardCharsets.UTF_8));
+      gzipOutputStream.close(); // close to flush and finish the compression
+      gzipData = byteArrayOutputStream.toByteArray();
+    }
+
+    // Mocking input stream with the gzip data
+    InputStream gzipStream = new ByteArrayInputStream(gzipData);
+
+    // Call the private method using reflection
+    InputStream resultStream = factory.createBasedOnEncodingHeader(gzipStream, encodingHeader);
+
+    // Decompress and validate the data matches original
+    ByteArrayOutputStream decompressedOutput = new ByteArrayOutputStream();
+    byte[] buffer = new byte[1024];
+    int bytesRead;
+    try (GZIPInputStream gzipInputStream = (GZIPInputStream) resultStream) {
+      while ((bytesRead = gzipInputStream.read(buffer)) != -1) {
+        decompressedOutput.write(buffer, 0, bytesRead);
```

Comment:
I think we shouldn't use write like this - if we call this multiple times I believe we would overwrite the beginning of the buffer. This shouldn't happen in this test, but better to do it properly either way